File: 343-integer-break/343-integer-break.py
class Solution:
    def integerBreak(self, n: int) -> int:
        
        @lru_cache(None)
        def helper(n):
            if n == 2:
                return 1
            elif n == 3:
                return 2
            
            maxx = 1
            for num1 in range(1,n//2 + 1):
                num2 = n - num1
                break_1 = num1 if num1 <= 3 else helper(num1)
                break_2 = num2 if num2 <= 3 else helper(num2)
                maxx = max(maxx,break_1*break_2)
            return maxx
        
        return helper(n)

File: 343-integer-break/343-integer-break.js
/**
 * @param {number} n
 * @return {number}
 */
var integerBreak = function(n) {

    const ans = {};
    const fetchCached = n => {
        if (!(n in ans)){
            ans[n] = helper(n);
        }
        return ans[n];  
    };
    
    const helper = (n) => {
        let maxPossible = 1;
        let num2 = 0;
        let break1 = 0;
        let break2 = 0;
        
        
        if (n in ans){
            return ans[n];
        }
        else if (n === 2 || n === 3){
            return n-1;
        }
        
        for(let num1 = 1;num1 <= Math.floor(n/2);num1++){
            num2 = n - num1;
            break1 = (num1 <= 3)? num1:fetchCached(num1);
            break2 = (num2 <= 3)? num2:fetchCached(num2);
            maxPossible = Math.max(maxPossible,break1 * break2);
        }
        ans[n] = maxPossible;
        return ans[n];
    };
    
    return helper(n);
    
};
